Refactor

savetheclocktower opened the pull request:
I have called this “refactor“ because I made a bunch of changes and they have no common theme.

* `ejs.render` does not appear to be an async function, so we can clean up a lot of `awaits` in `json2html.js`. (If there's an async version of `ejs.render` and we start using it, I can add them back in.)
* With some tinkering, I was able to get Less to regenerate stylesheets whenever a change is made, though I still want to figure out how to get it to regenerate _only_ the stylesheets so that I can see the changes after one second instead of ten.
* I've added a lot of the CSS changes you seemed to like.
* I moved the layouts in `pulsar-api/layouts` into `layouts/api` so that all the layouts can live in one directory. I was having trouble getting Eleventy to pick up on changes I made to those layout files, and this fixed that problem.
* No more generated content `{` and `}` to mark hovercard targets.
* Replaced the “blocks” layout with a flat list in at least one place, and I intend to make that change elsewhere as well.

I'm happy to keep this as a draft for a while as I continue to make changes:

* ~~I prefer the light mode theme and you appear to prefer the dark mode theme, but my goal is to make the colors in each theme resemble those on the main web site more closely.~~ (Done, more or less.)
* ~~On page load, the dark mode theme flashes for a second before it restores my light-mode preference, which is annoying.~~ (Fixed.)
* ~~At some point, I will be overriding so much of Skeleton's defaults that it might be worth discussing removing our dependency on it, though we're not at that point yet.~~ (Ehh, everything that's worth overriding has been overridden, and everything that's left would need its own replacement if we got rid of Skeleton. I don't plan to use any of its utility class names, but it can stay all the same.)
* ~~Hovercard triggers should be links. It's good to offer the contextual information when someone hovers, but they should also expect to click on the trigger and be taken to the appropriate page.~~ ~~(They are now links, but their `href`s are `#` until the card loads, at which point I fill them with the correct `href`. The long-term solution is to figure out how to bake that `href` into the source code directly — which may require more than one pass in the parsing phase.)~~ (This is now done, and it works in one pass because of various tricks and inferences we can make. (`foo-bar` is a package name, `FooBar` is a class name, and `::fooBar` and `Foo::bar` are unambiguous references to API methods in context.))
* ~~It would be useful to have some sort of convention for naming things in the CSS. BEM is my least favorite of these schemes… except for all the others. This is not urgent, though.~~ (This is done.)

savetheclocktower committed:
These changes seem to get eleventy to rebuild properly…

…when we change `.less` files.

(Still would be nice to have it know that it doesn't need to rebuild the entire project on style changes! Extra credit for later.)

savetheclocktower committed:
Changes to API templates

savetheclocktower committed:
Tweak some theme colors to look closer to those of the main Pulsar site

savetheclocktower committed:
Undo the “centered headings” override

savetheclocktower committed:
Change lots of global styles, including font family and heading sizes

savetheclocktower committed:
Styles for some API-page-specific markup

savetheclocktower committed:
More changes

savetheclocktower committed:
Point to new layout directories and remove stray `async`

savetheclocktower committed:
Add `.DS_Store` to `.gitignore`

savetheclocktower committed:
Turn hovercard triggers into links…

…with the eventual goal that they be linked to the relevant sections of the docs.

Also load the user's preferred theme as early as possible in the page lifecycle to prevent a flash of content.

savetheclocktower committed:
Use Less mixins to prevent duplication of dark mode variables

savetheclocktower committed:
Highlight the “active” heading in the sidebar…

…though it’s annoyingly complicated and ripe for simplification later.

savetheclocktower committed:
Whoops…

…should’ve added this a long time ago.

savetheclocktower committed:
Enforce minimum sidebar width

savetheclocktower committed:
Force long inline `code` blocks to break to the next line

savetheclocktower committed:
DRY the API building/writing flow

savetheclocktower committed:
Generate better HTML for custom containers

savetheclocktower committed:
Prevent images from overflowing the main container

savetheclocktower committed:
Add platform switcher

savetheclocktower committed:
Rewrite docs so that we utilize the new platform switcher…

…and show people only one platform's information (keybindings, file locations, menu items) at a time.

The switcher stays on the top of the screen via `position: sticky`, but we could easily switch this to `position: static` and keep it at the top of the page.

savetheclocktower committed:
Tweak some theme colors related to custom containers…

…which I'm still experimenting with.

savetheclocktower committed:
Rewrite all links in `11tydata.json` to be absolute…

…which I don't see a way around. The `../` trick in sidebar links wasn't going to work for anchor-style links anyway. This lets us render an item in the sidebar without caring about path relativity.

savetheclocktower commented:
Big new feature: the platform switcher! This was my favorite feature of the old docs.

<p><img width="638" alt="Screenshot 2024-03-24 at 12 37 06 PM" src="https://github.com/confused-Techie/pulsar-documentation-eleventy/assets/3450/c449a14b-8f37-4e73-bc46-b30511174960"></p>

<p><img width="631" alt="Screenshot 2024-03-24 at 12 37 16 PM" src="https://github.com/confused-Techie/pulsar-documentation-eleventy/assets/3450/d2311df0-aeb8-47db-ae27-e598b06041a0"></p>

Rewrote all the references to keybindings/menu items/file locations to use `span`s with inline class names for each platform, just like how the Flight Manual used to do it.

savetheclocktower committed:
Fix width on main content area

savetheclocktower committed:
Small docs fixes

savetheclocktower committed:
Make it possible to enforce a max width for all content…

…while keeping all sections aligned.

savetheclocktower committed:
Flesh out theming of custom containers…

…and give `kbd`s their own border color.

savetheclocktower committed:
Write some more `snippets` documentation

(long overdue!)

savetheclocktower committed:
Fix sticky positioning of platform switcher

savetheclocktower committed:
Restyle tab container to match new custom container styles

savetheclocktower committed:
Allow the platform selector to control platform-specific tab containers

savetheclocktower commented:
Restyled the custom containers:

<img src="https://github.com/confused-Techie/pulsar-documentation-eleventy/assets/3450/81bab634-cfc2-451f-8c84-4d4161739b92" width="300">
<img src="https://github.com/confused-Techie/pulsar-documentation-eleventy/assets/3450/677fe4ea-ff0c-4791-ad2f-58d8c35d2e55" width="300">

Also styled the tab containers to match:

<img width="826" alt="Screenshot 2024-03-24 at 4 26 30 PM" src="https://github.com/confused-Techie/pulsar-documentation-eleventy/assets/3450/bd00146b-98fe-4a79-b60d-d1f6f03676b8">

savetheclocktower committed:
Style the page header to resemble that of the main site…

…much more closely, at least.

savetheclocktower committed:
Make the theme switcher cycle between dark/light/auto

savetheclocktower committed:
Undo all letter-spacing tweaks on headings…

…to better match the main Pulsar site.

savetheclocktower committed:
A couple of docs fixes

savetheclocktower committed:
Update custom container colors to be closer to those on the website

savetheclocktower committed:
Custom container style tweaks

savetheclocktower committed:
Remove some tab containers…

…because they didn't play well with keyboard shortcuts with `[` and `]` in them. (Even when I used `kbd` tags! Even when I HTML-encoded the brackets!)

confused-Techie commented:
@savetheclocktower Thanks a ton for this!
I'll see what I can do about reviewing this in the near future, but at a glance these changes are looking awesome!

savetheclocktower commented:
Yeah, sorry. I started moving _really_ quickly and this PR got unwieldy.

savetheclocktower committed:
Make the generic “Summary” sidebar link point to the actual article name

savetheclocktower committed:
Write an “auto-TOC” module for the sidebar…

…to roughly match what we have on our blog.

Also reorganize `main.js`.

savetheclocktower committed:
`ol` style tweak

savetheclocktower committed:
Make dark-mode links the same color as light-mode links…

…just like on the main site

savetheclocktower committed:
Small docs tweak

savetheclocktower committed:
Cleanup sidebar code

savetheclocktower committed:
Make platform-switcher background match that of the header

savetheclocktower committed:
Fix AutoTOC for the last heading on a page

savetheclocktower committed:
Use ellipsis for AutoTOC entries to keep them to a single line

confused-Techie commented:
One thing I'll quickly add, from looking at these changes it looks like you've converted all of `[[Key]]` to `<kbd>Key</kbd>`.
I had switched those over thinking it simplified reading things and used `markdown-it-kbd` to handle the conversion, but if we would rather just use the actually HTML, you can probably also remove that package from the repo

savetheclocktower commented:
> One thing I'll quickly add, from looking at these changes it looks like you've converted all of `[[Key]]` to `<kbd>Key</kbd>`. I had switched those over thinking it simplified reading things and used `markdown-it-kbd` to handle the conversion, but if we would rather just use the actually HTML, you can probably also remove that package from the repo

I didn't convert them all — only in cases where the hotkey varies per platform. Otherwise I'd have to wrap `[[Key]]` in a tag like `<span class="platform-mac">` and I didn't see the point in that. Lots of the `[[Key]]` usages remain.

savetheclocktower committed:
Set the `meta[name="color-scheme"]` value to match the theme

savetheclocktower committed:
Rework `platform-switcher.less` to be more BEM

savetheclocktower committed:
Organize the CSS a bit better

savetheclocktower committed:
Fix some page header sizing issues

savetheclocktower committed:
Reorganize `sidebar.less`

savetheclocktower committed:
Nudge the active TOC link into view if it's needed

savetheclocktower committed:
Add next/previous links at bottom of manual pages

savetheclocktower committed:
Attach the helpers to `globalThis` to make linting happy

savetheclocktower committed:
Compare URLs for next/previous, not article titles

savetheclocktower committed:
Decrease tab size in `pre` blocks

savetheclocktower commented:
Maybe it would help if I gave a summary of my changes? Let me try to list them in rough order of impact on the codebase:

1. Addition of the platform switcher. Small impact for the UI of the switcher, large impact in that I replaced all instances of platform-specific keys (the **LNX:**/**MAC:**/**WIN:** sections). That’s probably the biggest reason why this PR touches so many files. The platform switcher starts out by making an educated guess based on your platform, but will remember your choice in `localStorage` the same way it remembers your theme.
2. As I had occasion to touch the Markdown files anyway, I made some changes to phrasing and content… but mostly I undid the hard-wrapping at 80 characters. I think that’s a holdover from the Flight Manual days, but it’s such a strange choice. I could’ve spent twice as much time reformatting each change I made to keep the hard wrap, but instead I just unwrapped each paragraph I touched. (If we want to keep the hard wrap, it should be applied automatically somehow, like a version of [Prettier](https://prettier.io/) but for Markdown. That must exist, right?)
3. I redid the sidebar metadata so that the `link` property didn’t just point to the article slug, but to the entire absolute path from the root. This means that the sidebar links can be used much more flexibly. I also added an `indexTitle` property so that the sidebar could display proper breadcrumbs rather than just use “Summary” for a page’s parent article.
4. The theme switcher. I wanted equivalent functionality to what we see on the main Pulsar site, so I borrowed the icons and the three-phase dark/light/auto behavior. It’s in auto mode by default.
5. The auto-TOC behavior is inspired by (but entirely reverse-engineered from) the similar behavior on the Pulsar site.
6. I made some changes to the layout. The sidebar is a fixed width and a permanent presence on the left side once the window is at least 1000px wide (we might be able to make room for it at a smaller breakpoint, too).
7. I restyled custom containers to resemble their equivalents on the main site as much as possible.
8. I changed our approach to “sibling” page navigation so that pages link to their next/previous siblings in the footer. This frees up more space in the sidebar for heading-based table-of-contents navigation.
9. Reworked the page header to function very similarly to that of the main site, including the underline effect on hover and the blurred background.

I’ve got some more changes in the pipeline, but **I don’t plan to put them into this PR** just because it’s already overloaded:

* I want to do some tweaking to the main page. That would involve adding a summary to each of the top-level sections, but that brings me to my next problem:

* I’d like to do some reorganizing of content so that we can eliminate at least one of...

    * Developing for Pulsar
    * Behind Pulsar
    * Package Guides

  …because I don’t know how to structure them or summarize each section’s content. They overlap in subject matter a bit too much.

* I imagine you’ve got thoughts about syntax highlighting, but I installed the [official syntax highlighting plugin](https://www.11ty.dev/docs/plugins/syntaxhighlight/) and it seems to be working fine. I gave it about half a dozen colors in each theme so that dark mode’s code blocks look like One Dark and light mode’s code blocks look like One Light. (It’s using [Prism](https://prismjs.com/) under the hood, which is pretty easy to understand… so I might even be able to write a syntax highlighter for SCM files!)

I know that I threw a lot at you, so if you’d rather see any of these changes in isolation, I’ll see if I can make it happen.

savetheclocktower committed:
Add syntax highlighting, fix submodule name, include full 1.114 JSON

savetheclocktower committed:
Overhaul hovercards…

* Make them resolve to links with proper `href`s.
* Preserve context for links like `{::onDidChange}}`.

savetheclocktower committed:
Make static hovercards link to MDN

savetheclocktower committed:
Fix adjacent nav for first article in a list

savetheclocktower committed:
Rename “Package API” in header to “Package Registry API”

savetheclocktower committed:
Make proper sidebar breadcrumbs for API pages

savetheclocktower committed:
Hovercards should contain summaries, not descriptions

savetheclocktower committed:
API pages shouldn't set URL hashes automatically

savetheclocktower committed:
Add Octicons

savetheclocktower committed:
Make dark mode header links the same color as on the main site

savetheclocktower committed:
Make the sidebar the sidebar at 750px rather than 100px

savetheclocktower committed:
More breathing room for page header links

savetheclocktower committed:
When the sidebar is not present, the main content area can spread out

savetheclocktower committed:
Harmonize borders on API listings, try a thick separator between methods

savetheclocktower committed:
Auto-detect package names and link them to the package registry

savetheclocktower committed:
Rearrange the order of generation…

…because there are some name collisions and something with a certain name in Pulsar should always beat something with the same name in another submodule. Hence Pulsar’s docs should be generated last.

savetheclocktower committed:
Syntax highlighting styling tweaks

savetheclocktower committed:
Fix links to API documentation from the manual

savetheclocktower committed:
API page tweaks

savetheclocktower committed:
Embrace the `skeleton-less` conventions a bit more

savetheclocktower committed:
Make the active item in the sidebar slightly bolder

confused-Techie commented:
Sorry I didn't respond earlier.

But I do agree that some of those sections are a bit strange. Since I've already restructured all of the content here, there were some pieces that didn't quite fit in the end.

But for the following here's what I'd think:
* Package Guides: I'd really hope we can keep this one, since I think it's monumentally helpful to have working examples and explanations of real code to learn things. And the package guides do exactly that. Plus that's somewhere I'd really like to expand on. As for summarizing, I'd go with something like "Example package's using common Pulsar Features"
* Developing for Pulsar: I honestly think this one is fine as is, being the most fleshed out of all the sections you called out
* Behind Pulsar: This is absolutely one that doesn't fit. Maybe we could pull that section apart putting much of it's information into "Using Pulsar" as that's where we dive into specific features, and this could work there? At least for things like Keymaps, whereas serialization and such can be moved into the developing section.

What do you think of the above?

Also now that I have some time I'm going to get on testing all the changes in this PR, so again my bad on taking a while, but it seems you've certainly found lots to improve in the meantime

confused-Techie commented:
Also @savetheclocktower Just wanting to say, I've finally fired up all of these changes locally, and just wow. It looks amazing, I'm very glad you've opted to leave much of the hovercard behavior since I was a fan of that, but you are right that with the right styling the list of items does look great on overview pages, and I love that you brough over the auto-TOC from the other site, that one does really pay off.

The portability of everything is quite impressive, making sure that hovercard resolution within the API docs resolve not only to current versions but also classes, and the changes to links in the sidebar making things a bit easier to work with is all awesome. I'll poke around a bit more then look at the code and see if we can get this one merged before it becomes too unwieldy

savetheclocktower commented:
Great, thanks.

As for restructuring, I've been saving that for another PR, but I'll come up with something. I don't want it to seem like something you have to read all of, but I would like for it to make sense _if_ you wanted to read it in linear fashion.

confused-Techie commented on eleventy.config.js:
If we are adding the `less` handling right in the config, do we want to go ahead and get rid of `./plugins/less.js`?

savetheclocktower replied:
Good point. I'll do that.

(I moved that into this config file because I needed access to `this.addDependencies` and it's not clear to me that there's a way to inject that scope into a plugin defined in another file.)

confused-Techie replied:
Just saw the commit for this, awesome!

confused-Techie commented on layouts/page_header.ejs:
While I'm in favour of using the raw SVG data instead of linking to the image, I am very curious on where you got the images from if you recall. Just to know for myself and if we ever need to change them later on

savetheclocktower replied:
Do you mean the Pulsar logos? I'm pretty sure I grabbed the files from the main website.

I definitely grabbed the SVGs for the theme switcher from the website.

confused-Techie replied:
Ah yeah I was referring to the theme switcher logos. So good deal then that sounds fine by me

confused-Techie commented on markdown-it-plugins/hovercard.js:
I am a huge fan of how you fit this into the actual Hovercard parsing itself. Like you've already done this allows us contextual references which are so valuable. 

Although I do wonder if you have any thoughts on adding the fully hovercard resolution here, so that we can add all the other information about the hovercards into `data-` attributes of the element, that way there's no client side JavaScript API requests for further details. Might be worth it to add this here, but curious if that's something you looked at at all.

(To be clear absolutely not something that needs to be implemented on this PR, just curious)

savetheclocktower replied:
> I am a huge fan of how you fit this into the actual Hovercard parsing itself. Like you've already done this allows us contextual references which are so valuable.

It's serendipity that the various things we link to can be told apart just by inspecting the string and its casing conventions. I was a bit surprised I got it to work myself. (Though it's useful to know that Markdown-IT plugins have access to “environment” data; we can probably get some mileage out of that in the future.)

> Although I do wonder if you have any thoughts on adding the fully hovercard resolution here, so that we can add all the other information about the hovercards into `data-` attributes of the element, that way there's no client side JavaScript API requests for further details. Might be worth it to add this here, but curious if that's something you looked at at all.

Yeah, I think those are good ideas for future enhancements.

An incremental step might be to bake the hovercard data into a single JSON file rather than having one for each token. Then that data could be `fetch`ed once… or perhaps even baked into the page source.

confused-Techie replied:
Ahh interesting idea for an incremental improvement. I'll add it as an issue to this repo so I don't forget

confused-Techie reviewed (approved):
Alright, finally took a look at this one.

Since you famously have really fantastic code I didn't worry to much about the specifics here, mostly reviewing in broad strokes.

Which all in all things look fantastic! Trying out the site locally I'm a huge fan of how it works and feels. The changes made to the hovercards, both in presentation (with the href attribute) and resolution (contextual!) I'm a huge fan. As for the CSS changes there's a lot, but really the site doesn't stray to a point where I feel like it's difficult or contradicts what this revision set out to do, so I'm totally on board. The only slight annoyance is the auto-generated TOC getting cut off on some resolutions. If anything we may want to check about adding a title tag to reveal the full text on hover. 

I've added a couple other small notes, and will of course give you a chance to respond, but overall this is awesome, and I'd be happy to merge it

savetheclocktower committed:
Delete `plugins/less.js`

savetheclocktower commented:
> The only slight annoyance is the auto-generated TOC getting cut off on some resolutions. If anything we may want to check about adding a title tag to reveal the full text on hover.

Yeah, good call. I'll add a `title` attribute to all generated links later on.
